circuit/src/merkle_tree/native.rs

```diff
@@ -0,0 +1,173 @@
+use std::marker::PhantomData;
+
+use halo2wrong::halo2::arithmetic::FieldExt;
+use num_traits::pow;
+use std::collections::HashMap;
+
+use crate::{params::RoundParams, poseidon::native::Poseidon};
+const WIDTH: usize = 5;
+
+#[derive(Clone, Debug)]
+/// MerkleTree structure
+struct MerkleTree<F: FieldExt, P>
+where
+	P: RoundParams<F, WIDTH>,
+{
+	/// PhantomData for the params
+	_params: PhantomData<P>,
+	/// Variables with level and value to represent nodes
+	nodes: HashMap<usize, Vec<F>>,
+	/// Height of the tree
+	height: usize,
+}
+
+impl<F: FieldExt, P> MerkleTree<F, P>
+where
+	P: RoundParams<F, WIDTH>,
+{
+	/// Create a new empty Merkle Tree with given height
+	fn new(height: usize) -> Self {
+		// 0th level is the leaf level and the max level is the root level
+		let mut nodes = HashMap::new();
+		for i in 0..height + 1 {
+			let num_nodes = pow(2, height - i);
+			let nodes_vec = vec![F::zero(); num_nodes];
+			nodes.insert(i, nodes_vec);
```

Comment:
While leaf nodes can be zero, it doesnt make any sense for upper level nodes to be zero, they should always be hashes of 2 child nodes.

I think we should merge all three functions into one, such that:
You pass `height` as the height of the tree, and `leaves` as leaves of the tree, but the length of leaves array should be <= 2 ** height. If it's less than that, the rest of the leaves should be defaulted to zero, so that the leaves array is filled to the next power of two.

Reply:
I also took out height variable from function inputs. I did height calculations inside the circuit so we don't have to concern about if number of leaves is bigger than the 2**height.

Reply:
I think in our use-case we want to explicitly pass the `height`, since it will be predetermined inside a smart contract or similar. In case of leaves array being bigger than height, we can panic, by adding `assert!(leaves.len() <= 2**height)` or similar